crates/ruff_linter/src/linter.rs

```diff
@@ -260,6 +260,29 @@ pub fn check_path(
         }
     }
 
+    // Update fix applicability to account for overrides
+    if !settings.extend_safe_fixes.is_empty() || !settings.extend_unsafe_fixes.is_empty() {
+        for diagnostic in &mut diagnostics {
+            if let Some(fix) = diagnostic.fix.take() {
+                // Check unsafe before safe so if someone puts a rule in both we are conservative
+                if settings
+                    .extend_unsafe_fixes
+                    .contains(diagnostic.kind.rule())
+                    && fix.applicability().is_safe()
+                {
+                    diagnostic.set_fix(fix.with_applicability(Applicability::Unsafe));
+                } else if settings.extend_safe_fixes.contains(diagnostic.kind.rule())
+                    && fix.applicability().is_unsafe()
+                {
+                    diagnostic.set_fix(fix.with_applicability(Applicability::Safe));
+                } else {
+                    // Retain the existing fix (will be dropped from `.take()` otherwise)
+                    diagnostic.set_fix(fix);
```

Comment:
I think you could also write this section like:

```rust
if let Some(fix) = &mut diagnostic.fix {
    // Check unsafe before safe so if someone puts a rule in both we are conservative
    if settings
        .extend_unsafe_fixes
        .contains(diagnostic.kind.rule())
        && fix.applicability().is_safe()
    {
        fix.applicability = Applicability::Unsafe;
    } else if settings.extend_safe_fixes.contains(diagnostic.kind.rule())
        && fix.applicability().is_unsafe()
    {
        fix.applicability = Applicability::Safe;
    }
}
```

If `applicability` is `pub(crate)`, or more likely with something like `set_applicability(&mut self, applicability: Applicability)` on `Fix`. But I think I prefer what you have.

Reply:
Thanks for explaining!

Reply:
To add one more alternative:
```rust
            diagnostic.fix = diagnostic.fix.take().map(|fix| {
                // Enforce demotions over promotions so if someone puts a rule in both we are conservative
                if fix.applicability().is_safe()
                    && settings
                        .extend_unsafe_fixes
                        .contains(diagnostic.kind.rule())
                {
                    fix.with_applicability(Applicability::Unsafe)
                } else if fix.applicability().is_unsafe()
                    && settings.extend_safe_fixes.contains(diagnostic.kind.rule())
                {
                    fix.with_applicability(Applicability::Safe)
                } else {
                    // Retain the existing fix (will be dropped from `.take()` otherwise)
                    fix
                }
            });
```
(it's annoying that the compiler doesn't accept this without `.take()`)